dashboard/app/controllers/api/v1/text_to_speech_controller.rb

```diff
@@ -1,18 +1,41 @@
 class Api::V1::TextToSpeechController < Api::V1::JsonApiController
   include AzureTextToSpeech
 
+  # Allowed number of unique requests per minute before that client is throttled.
+  # These values are fallbacks for DCDO-configured values used below.
+  REQUEST_LIMIT_PER_MIN_DEFAULT = 100
+  REQUEST_LIMIT_PER_MIN_IP = 1000
+
   # POST /dashboardapi/v1/text_to_speech/azure
   # @param [String] params[:text] Text to convert to speech
   # @param [String] params[:gender] Gender of voice
   # @param [String] params[:locale] I18n locale of voice. Optional. Uses request locale if not provided.
   # @returns [ArrayBuffer] Binary data that can be played as sound in the browser.
   def azure
-    speech_array_buffer = AzureTextToSpeech.get_speech(params[:text], params[:gender], locale)
-    if speech_array_buffer.nil_or_empty?
-      return head :bad_request
-    else
-      send_data speech_array_buffer, type: 'arraybuffer'
+    id = current_user&.id || session.id
+    # Only throttle by IP if no user or session ID is available.
+    throttle_ip = id.blank?
+    id ||= request.ip
+    limit = throttle_ip ?
+      DCDO.get('azure_tts_request_limit_per_min_ip', REQUEST_LIMIT_PER_MIN_IP) :
+      DCDO.get('azure_tts_request_limit_per_min_default', REQUEST_LIMIT_PER_MIN_DEFAULT)
+    period = 60
+
+    AzureTextToSpeech.throttled_get_speech(params[:text], params[:gender], locale, id, limit, period) do |speech_array_buffer|
+      if speech_array_buffer.nil_or_empty?
+        return head :bad_request
+      else
+        return send_data speech_array_buffer, type: 'arraybuffer'
+      end
     end
+
+    # If we make it here, the request should be throttled.
```

Comment:
My lack of understanding of Ruby's `yield` had me confused here, at first, so thanks for walking me through all of this "in person".  I guess a newcomer like me might like to know that the `yield` calls inside `throttled_get_speech` will invoke the above block, and when the `speech_array_buffer` is `nil` then this block has the `azure` method return a `head :bad_request`, and when it has a valid `speech_array_buffer` then it sends that as expected.  If, on the other hand, `throttled_get_speech` discovers that throttling is needed, then it returns immediately without doing a `yield`, and so the above block is never executed, and so the `Honeybadger` notification just below happens.  It's also interesting that the `return yield` calls in `throttled_get_speech` simply ensure that the function both calls the above block and also returns immediately after doing that, though the idea of returning the value returned by the block doesn't really happen; rather, the block's return value is what goes directly back as the return value of `azure`.

Not sure if worth adding some comments to tell this story, but at any rate wanted to capture what I learned here.